### ISSUE TITLE
Using router helpers during compilation time throws an ets lookup error

### ISSUE THREAD

**Issue by lessless** [phoenixframework/phoenix#2841]
### Environment

* Elixir version (elixir -v): 1.6.0 (compiled with OTP 20)
* Phoenix version (mix deps):  1.3.2
* Operating system: OS X

### Expected behavior
It's possible to pass a `path` as an argument to the plug

### Actual behavior
Using router helper during compilation throws an ETS lookup error

controller:

```
defmodule PathCacheWeb.PageController do
  use PathCacheWeb, :controller
  plug PathCacheWeb.Plug.DummyPlug, [redirect_path: page_path(PathCacheWeb.Endpoint, :index)]
```

backtrace: 
```
= Compilation error in file lib/path_cache_web/controllers/page_controller.ex ==
** (ArgumentError) argument error
    (stdlib) :ets.lookup(PathCacheWeb.Endpoint, :__phoenix_path__)
    (phoenix) lib/phoenix/config.ex:45: Phoenix.Config.cache/3
    lib/phoenix/endpoint.ex:643: PathCacheWeb.Endpoint.path/1
    lib/path_cache_web/controllers/page_controller.ex:6: (module)
    (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
    (elixir) lib/kernel/parallel_compiler.ex:198: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
```

I suspect that it’s because of the `:ets.lookup` tries to access non-existing table which is being created only in the runtime https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/config.ex#L102

Sample repository https://github.com/lessless/path_cache_bug

P.S.  simpler way to reproduce is just to add a module attribute

```
defmodule PathCacheWeb.PageController do
  use PathCacheWeb, :controller
  @lolo page_path(PathCacheWeb.Endpoint, :index)
```

**Comment by chrismccord** [comment]
It would be very difficult for us to provide this because the application can't be started during compilation... because we are compiling it :). Also remember that plug `init`'s are compiled time invocations. We/you could try to force the endpoint to start during compilation, but this is also tricky because it may want to start the web server depending on your configuration. So the best solution to this is to pass an mfa to the plug and evaluation it at runtime, for example:

```elixir
defmodule PathCacheWeb.PageController do
  use PathCacheWeb, :controller
  plug SomePlug, 
    redirect_path: {Routes, :page_path, [PathCacheWeb.Endpoint, :index]}

...

defmodule SomePlug do

  def init(opts), do: Enum.into(opts), %{})

  def call(conn, %{redirect_path: {router, func, args}) do
    case ... do
      ... -> Phoenix.Controller.redirect(conn, to: apply(router, func, args))
    end
  end
end
```

I will improve the ets argument error to let the user know the endpoint isn't running to make things clearer, but either a captured named function or passing an mfa to the plug is the way to go. Hope that helps!

**Comment by josevalim** [comment]
I talked to @chrismccord and I think we could try to make it work by simply disabling the cache if the ets table is not there. We will give it a try, but not guarantees yet.

**Comment by lessless** [comment]
Thanks guys! getting around that was pretty easy, but I imagine it can be a source of frustration for somebody else. 

@josevalim see you next week in Taipei! ;)

**Comment by msudgh** [comment]
I have the same error while getting `Endpoint.url()`

```ex
def get_default_avatar() do
  Path.join([Endpoint.url(), Application.get_env(:blinder, :static_path), @default_avatar_file_name])
end
```

```bash
== Compilation error in file lib/blinder/patrons/patron.ex ==
** (ArgumentError) argument error
    (stdlib) :ets.lookup(TestApi.Endpoint, :__phoenix_url__)
    (phoenix) lib/phoenix/config.ex:42: Phoenix.Config.cache/3
    lib/util.ex:13: TestApi.Util.get_default_avatar/0
    lib/test/patrons/patron.ex:17: (module)
    (stdlib) erl_eval.erl:680: :erl_eval.do_apply/6
    (elixir) lib/kernel/parallel_compiler.ex:208: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
```

what could be wrong?

**Comment by chrismccord** [comment]
If you aren't calling it at compile time, then it would imply your endpoint is not started in your sup tree

**Comment by TheFirstAvenger** [comment]
@josevalim I took a swing at implementing the "skip the cache if the ets table is not there" route, however unless I am missing something, we still need the ets table live because the function that is called on a cache miss itself is looking in the ets table for that config. For example, when calling a helper function like this (modeled after the original posts example):

```
defmodule MyappWeb.Login do
  use MyappWeb, :controller
  plug MyappWeb.Plugs.DummyPlug, [to: Routes.login_path(MyappWeb.Endpoint, :index)]
```

with the below change to config.ex:

```elixir
  def cache(module, key, fun) do
    try do
      case :ets.lookup(module, key) do
        [{^key, :cache, val}] ->
          val

        [] ->
          case fun.(module) do
            {:cache, val} ->
              :ets.insert(module, {key, :cache, val})
              val

            {:nocache, val} ->
              val
          end
      end
    rescue
      e in ArgumentError ->
        case :ets.info(module) do
          :undefined ->
            {_, val} = fun.(module)
            val

          _ ->
            raise e
        end
    end
  end
```

it still results in an `:ets` error on compilation:

```
** (ArgumentError) argument error
    (stdlib 3.12.1) :ets.lookup(MyappWeb.Endpoint, :url)
    lib/phoenix/endpoint.ex:493: MyappWeb.Endpoint.config/2
    (phoenix 1.6.0-dev) lib/phoenix/endpoint/supervisor.ex:255: Phoenix.Endpoint.Supervisor.path/1
    (phoenix 1.6.0-dev) lib/phoenix/config.ex:68: Phoenix.Config.cache/3
    lib/phoenix/endpoint.ex:555: MyappWeb.Endpoint.path/1
    lib/myapp_web/controllers/login.ex:3: (module)
```

only this time it is complaining because of this line:

https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/supervisor.ex#L255

which calls this config lookup in `:ets`:

https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint.ex#L492-L497

**Comment by josevalim** [comment]
@TheFirstAvenger thanks for looking into this. perhaps the best course of action here is to catch the ets calls and raise with a clearer error message:

```
try do
  :ets...
rescue
  _ -> raise ArgumentError, "could not find ets table for endpoint #{...}. Make sure your endpoint is started and note you cannot access endpoint functions at compile-time"
end
```

**Comment by TheFirstAvenger** [comment]
@josevalim Implemented in #4240 as suggested
